### DISCUSSION TITLE
Multiple Archaeologists per Sarcophagus

### DISCUSSION THREAD

**Discussion by adamgall** [sarcophagus-org/sarcophagus-v2-specs#2] [Ideas]
# What's the problem

The most common piece of critical feedback about version 1 was the trust / bottleneck / point of failure in the fact that a single archeologist (which corresponds to a single service running on someone's server) was responsible for being online, active, and able to complete their duties as an archeologist at some undefined point in "the future". And if they _weren't_, everything kinda fell apart.

In the case of a bad-actor archaeologist:

* That single archaeologist holds the key to unwrap the outer layer of a sarcophagus, which breaks the point of a dead-man's switch system (don't reveal information until the switch is triggered).

In the case of an unresponsive archaeologist at the time of the switch triggering:

* Embalmers didn't receive the service they paid for.
* Recipients couldn't get access to the data they were expecting.
* The archaeologist themself didn't get paid (but this is certainly understandable, it just sucks).

# Let's fix it

One major goal in Sarcophagus V2 is to support the idea that _many_ archaeologists can be associated with a single sarcophagus. This has the following positive effects:

* The "outer layer" encryption keys for a sarcophagus are actually split up amongst _many_ archaeologists, removing the risk that any one archaeologist has all of the information needed to reveal the contents of the sarcophagus.
* At the time of the dead man's switch triggering, any set of `m` archaeologists of the full set `n` archaeologists can come together to re-form the key to the sarcophagus, which allows much greater resiliency in the system if some archaeologists are offline or unresponsive at the time of unwrapping.

This discussion exists to work through the implementation details necessary to bring this exciting feature to life!

**Comment by sethhrbek** [comment]
#PROBLEM
If an embalmer is going to select multiple archs on the front-end, how do we make this process somewhat seamless?

Trying to identify 5 archaeologists (potentially more in future) could be rather laborious for the end-user.

David mentioned there had been discussion of a "lead" archaeologist that the embalmer would select, and then the Sarcophagus system could intelligently pick the remaining archs.

We could do a "suggest" button on the front end which uses various stats to determine best archs to choose. So the user could select 0 or up to 5 (or whatever max is),

**Reply by adamgall** [reply]
First point of clarification, to make sure we are all on the same page about the right level of abstraction here:

Do the contracts themselves care about, or enforce any rules about, which archaeologists are chosen for a sarcophagus?

Or is the problem being discussed here just at the layer of the webapp?

My gut says that the contracts shouldn't try to enforce any specific behavior around which archaeologists are being selected on a new sarcophagus, _except_ to confirm that they're all unique.

**Reply by sethhrbek** [reply]
We were viewing this as a web-app UX/UI issue, not contract level (which is why I had initially separated into a separate issue).

**Reply by adamgall** [reply]
> We were viewing this as a web-app UX/UI issue, not contract level (which is why I had initially separated into a separate issue).

Gotcha, understood. To be clear I didn't necessarily create this discussion _just_ from the context of the contracts, but more so about the feature holistically. I do totally expect that as we dig into all of this, it'll result in many new discussions and issues in more appropriate repositories.

**Comment by sethhrbek** [comment]
Who is generating the initial private key for the outer-layer of encryption (the key that will be sharded)? Is it the embalmer? Is there any security concerns around the embalmer somehow intercepting this private key during the encryption process if the p-key is being generated client-side?

**Reply by adamgall** [reply]
My thought was that yes, it must be the embalmer that is generating the new keypair that'll be used on the sarcophagus that they're creating.

Can you explain the security concerns you're worried about with a little more detail? The concept of "generating private keys client side" is a very common pattern...

**Reply by sethhrbek** [reply]
My general assumption was that generating private keys client side means the one generating client-side has access to them (I'm generating them for myself). But this may be a very wrong assumption based on my anecdotal experience generating them client-side. 

In this case we don't want the embalmer to know the private key, even though they are the one generating it. I have *no* idea how the embalmer would intercept these private keys, I just wanted to make sure it was 100% that they can't access it.

**Reply by adamgall** [reply]
I mean, you're right. The client (browser) generating a private key could very much be intercepted by the human running the browser on their machine.

Even in the case where someone is avoiding the browser entirely, and interacting directly with the contracts and archaeologists, "the system" expects that they're generating the outer encryption key themself.

You're positing that this is a problem. Hmmmmmm.

It's a problem because... during the lifecycle of a sarcophagus, if the inner-encrypted layer gets "leaked", and proofed to the contracts, the contracts will penalize the archaeologists. In the original version, if the inner layer was leaked, then guaranteed that leak came from the archaeologist themself, because they were the only one with the private key necessary to get to that inner layer. But in this new model, a different actor (embalmer) technically could hold onto that private key, decrypt their own sarcophagus, submit it to the contracts, which would result in a bond slash for all the archaeologists.

**Reply by adamgall** [reply]
I agree this is a problem.

**Reply by adamgall** [reply]
Maybe since we now have multiple archaeologists per sarcophagus, we need to re-think our bonding/slashing behaviors.

**Reply by sethhrbek** [reply]
The embalmer has access to the unencrypted payload -- they are responsible for uploading this in v1. They could just as easily release this payload early as they could the private key --- so maybe it's ok to trust them with the PK since we have to trust them anyways?

**Reply by adamgall** [reply]
Trying to think this through...

In V1 -- `accuseArchaeologist ` can be called by anyone who has access to the "inner" encrypted payload. that is, the data blob that should only be accessible to the archaeologist (because they and only they hold the key to decrypt the "outer" encrypted blob). This function exists to penalize the archaeologist, if the archaeologist leaked their key or the "inner" encrypted payload early.

The assumption here is that ONLY the archaeologist know the "inner" encrypted blob. But that's not the case. In V1, the embalmer (who knows the plaintext payload, obviously) can always generate that "inner" encrypted blob any time they want, by re-encrypting it with the archaeologist's public key.

In V2, although the workflow is different, I don't think there are any new security implications. In this new model in which the embalmer generates the pub/priv keypair, they are still able to generate that "inner" encrypted blob.

@sethhrbek are we talking about the same thing here?

**Reply by Afoil5454** [reply]
I don't want to divert but wanted to add to the conversation for potential strategies

Would the potential for using MPC as an extension of SSS be an option? From my understanding, Its possible to add or multiply to the private key shards without revealing secrets allowing us to potentially compute arbitrary functions on the secrets

**Reply by adamgall** [reply]
@Afoil5454 that sounds pretty cool indeed. What does "MPC" stand for? I'm finding references to something called "Secure Multiparty Computation" when I google it.

**Reply by adamgall** [reply]
https://medium.com/wanchain-foundation/secure-multiparty-computation-and-shamirs-secret-sharing-on-wanchain-e502012b80ef

**Reply by Afoil5454** [reply]
correct, multi-party computation

Its a way of combining shards to compute the combined secret without actually revealing the secret shards

This should allow for easier UX as you wouldnt need to re-create new shards after the shards have been comined as opposed to the basic SSSS 

This would also allow for much more complex interactions with the archaeologists by the Embalmer

**Reply by sethhrbek** [reply]
> @sethhrbek are we talking about the same thing here?

Yes -- that is exactly my thought process. The result of this thought process being there are no additional security implications with the new model (specifically the embalmer generating the keys). Hence my previous concern regarding the embalmer generating these keys is null and void.
